README.md
# @keleran/routing

Utilities to simplify request routing in Node.js projects.  
Use ESM, require Node.js 16+.

![GitHub Workflow Status](https://img.shields.io/github/workflow/status/dchekanov/routing/Test)

## Installation

```bash
$ npm i @keleran/routing
```

```javascript
import {discover, mount} from '@keleran/routing';
```

## `discover`

Returns a flat list of route handlers found in the specified directory. 
If you don't need to work with the list before mounting, use `mount` directly - it returns the same data.

```javascript
// [{route: String, method: String, source: Object, pipeline: Array}, ...]
const handlers = await discover({dir: 'routes'});
```

## `mount`

Reads a file tree and mounts request handlers to the specified Express app.

Request handler is a file named as "HTTP_METHOD.js" (get.js, post.js, etc.) that exports an object
with the following properties:

1. "rateLimit" (optional) - a string or an object that defines rate limiting options 
for the [rate-limiter-flexible](https://github.com/animir/node-rate-limiter-flexible) library.  
The string is parsed with the [ms](https://github.com/vercel/ms) library, 
"1s" means "allow 1 request per second from this IP".  
The object is passed to rate-limiter-flexible as is. HTTP 429 will be thrown upon exceeding the limit.
2. "authorize" (optional) - an authorization function that MUST return true for request to be processed.
Otherwise it throws HTTP 403 using the [http-errors](https://github.com/jshttp/http-errors) library. 
Supplied with the "req" and "res" arguments. Can be async.
3. "middleware" (optional) - a function or an array of functions that are regular middleware 
with the "req", "res", and "next" arguments supplied.
4. "handle" - the main function that handles the request. Supplied with the "req" and "res" arguments. Can be async.

The order of execution matches the order listed above.

### Example

The following tree:

```
routes
├── =
│   └── get.js
└── route
    ├── get.js
    └── =param-name
        └── get.js
``` 

With the call:

```javascript
await mount({dir: 'routes', app, route: '/test'});
```
 
Will be mounted like this (notice the order and param/wildcard substitution):
 
```javascript
app.get('/test/route', /* handler defined in routes/route/get.js */);
app.get('/test/route/:paramName', /* handler defined in routes/route/=param-name/get.js */);
app.get('/test/*', /* handler defined in routes/=/get.js */);
```


CHANGELOG.md
# Changelog

## 4.0.0

* Dropped support for Node.js 14.
* Added support for Node.js 20.
* Updated dependencies.

## 3.0.0

* Converted to ESM.
* Dropped Node.js 12 support.
* Dropped SonarQube scanning.
* Invalid route file content triggers error with code "MODULE_INVALID" instead of "SOURCE_INVALID".
* Node.js 18 has been added to the test matrix.
* Updated dependencies.


test/index.js
import assert from 'assert';
import sinon from 'sinon';
import {discover, mount} from '../index.js';
import httpErrors from 'http-errors';

const {HttpError} = httpErrors;

/**
 * Simulate middleware pipeline processing.
 */
function pipe({pipeline, req, res, next} = {}) {
  async function runStage(idx) {
    let stop;

    if (!pipeline[idx]) {
      return;
    }

    await pipeline[idx](req, res, err => {
      next(err);

      if (err) {
        stop = true;
      }
    });

    if (!stop) {
      return runStage(idx + 1);
    }
  }

  return runStage(0);
}

describe('discover', function () {
  it('throws if the "dir" parameter is not a non-empty string', async function () {
    await assert.rejects(() => discover({}), {name: 'Error', code: 'DIR_INVALID'});
    await assert.rejects(() => discover({dir: true}), {name: 'Error', code: 'DIR_INVALID'});
    await assert.rejects(() => discover({dir: 10}), {name: 'Error', code: 'DIR_INVALID'});
    await assert.rejects(() => discover({dir: ''}), {name: 'Error', code: 'DIR_INVALID'});
  });

  it('throws if the "dir" parameter points to a non-existing directory', async function () {
    await assert.rejects(() => discover({dir: 'missing'}), {name: 'Error', code: 'ENOENT'});
  });

  it('throws if the "route" parameter is not a non-empty string', async function () {
    await assert.rejects(() => discover({dir: 'test/routes/plain', route: true}), {
      name: 'Error',
      code: 'ROOT_INVALID'
    });
    await assert.rejects(() => discover({dir: 'test/routes/plain', route: 10}), {name: 'Error', code: 'ROOT_INVALID'});
    await assert.rejects(() => discover({dir: 'test/routes/plain', route: ' '}), {name: 'Error', code: 'ROOT_INVALID'});
  });

  it('throws if route file does not export the "handle" function', async function () {
    await assert.rejects(
      () => discover({dir: 'test/routes/no-handle'}),
      {name: 'Error', code: 'HANDLE_INVALID'}
    );
  });

  it('throws if route file exports an invalid "rateLimit" member', async function () {
    await assert.rejects(
      () => discover({dir: 'test/routes/rate-limit-invalid'}),
      {name: 'Error', code: 'RATE_LIMIT_INVALID'}
    );
  });

  it('throws if route file exports an invalid "authorize" member', async function () {
    await assert.rejects(
      () => discover({dir: 'test/routes/authorize-invalid'}),
      {name: 'Error', code: 'AUTHORIZE_INVALID'}
    );
  });
});

describe('mount', function () {
  const app = {};

  ['get', 'post', 'put', 'patch', 'delete'].forEach(method => app[method] = sinon.fake());

  const req = sinon.fake();

  const res = {
    end: sinon.fake()
  };

  const next = sinon.fake();

  beforeEach(() => sinon.reset());

  it('throws when neither "dir" nor "handlers" is supplied', async function () {
    await assert.rejects(() => mount({}), {name: 'Error', code: 'MODULE_INVALID'});
  });

  it('mounts a plain list of routes', async function () {
    await mount({app, dir: 'test/routes/plain'});

    ['get', 'post', 'put', 'patch', 'delete'].forEach(method => {
      assert(app[method].calledOnce);
      assert(app[method].firstCall.args[0] === '/');
      assert(Array.isArray(app[method].firstCall.args[1]));
      assert(app[method].firstCall.args[1].every(element => typeof element === 'function'));
    });
  });

  it('mounts routes when supplied with already discovered handlers', async function () {
    const handlers = await discover({dir: 'test/routes/plain'});

    await mount({app, handlers});

    ['get', 'post', 'put', 'patch', 'delete'].forEach(method => {
      assert(app[method].calledOnce);
      assert(app[method].firstCall.args[0] === '/');
      assert(Array.isArray(app[method].firstCall.args[1]));
      assert(app[method].firstCall.args[1].every(element => typeof element === 'function'));
    });
  });

  it('mounts a tree of routes and returns a report', async function () {
    const mounted = await mount({app, dir: 'test/routes/tree'});

    assert(mounted.length === 2);
    assert(mounted[0].route === '/');
    assert(mounted[0].method === 'get');
    assert(mounted[0].module === await import('./routes/tree/get.js'));
    assert(Array.isArray(mounted[0].pipeline));

    assert(mounted[1].route === '/branch');
    assert(mounted[1].method === 'get');
    assert(mounted[1].module === await import('./routes/tree/branch/get.js'));
    assert(Array.isArray(mounted[1].pipeline));

    assert(app.get.calledTwice);
    assert(app.get.firstCall.args[0] === '/');
    assert(app.get.secondCall.args[0] === '/branch');
  });

  it('mounts dirs with names starting with "=" the last and replaces "=" with ":" in route path', async function () {
    await mount({app, dir: 'test/routes/params'});

    assert(app.get.calledTwice);
    assert(app.get.firstCall.args[0] === '/regular');
    assert(app.get.secondCall.args[0] === '/:paramNameValue');
  });

  it('mounts dirs with names equal to "=" the last and replaces "=" with "*" in route path', async function () {
    await mount({app, dir: 'test/routes/wildcard'});

    assert(app.get.calledTwice);
    assert(app.get.firstCall.args[0] === '/regular');
    assert(app.get.secondCall.args[0] === '/*');
  });

  it('mounts a pipeline that executes "handle"', async function () {
    await mount({app, dir: 'test/routes/pipeline/handle'});

    await pipe({pipeline: app.get.firstCall.args[1], req, res, next});

    assert(res.end.calledOnce);
    assert(res.end.firstCall.args[0] === 'ok');
  });

  it('mounts a pipeline that rate limits (string)', async function () {
    await mount({app, dir: 'test/routes/pipeline/rate-limit/string'});

    await pipe({pipeline: app.get.firstCall.args[1], req, res, next});
    await pipe({pipeline: app.get.firstCall.args[1], req, res, next});

    assert(res.end.calledOnce);
    assert(res.end.firstCall.args[0] === 'ok');

    assert(next.calledTwice);
    assert(next.secondCall.args[0] instanceof HttpError);
    assert(next.secondCall.args[0].status === 429);
  });

  it('mounts a pipeline that rate limits (object)', async function () {
    await mount({app, dir: 'test/routes/pipeline/rate-limit/object'});

    await pipe({pipeline: app.get.firstCall.args[1], req, res, next});
    await pipe({pipeline: app.get.firstCall.args[1], req, res, next});

    assert(res.end.calledOnce);
    assert(res.end.firstCall.args[0] === 'ok');

    assert(next.calledTwice);
    assert(next.secondCall.args[0] instanceof HttpError);
    assert(next.secondCall.args[0].status === 429);
  });

  it('mounts a pipeline that authorizes (success)', async function () {
    await mount({app, dir: 'test/routes/pipeline/authorize/success'});

    await pipe({pipeline: app.get.firstCall.args[1], req, res, next});

    assert(next.calledOnce);
    assert(typeof next.firstCall.args[0] === 'undefined');

    assert(res.end.calledOnce);
    assert(res.end.firstCall.args[0] === 'ok');
  });

  it('mounts a pipeline that authorizes (failure)', async function () {
    await mount({app, dir: 'test/routes/pipeline/authorize/failure'});

    await pipe({pipeline: app.get.firstCall.args[1], req, res, next});

    assert(next.calledOnce);
    assert(next.firstCall.args[0] instanceof HttpError);
    assert(next.firstCall.args[0].status === 403);
  });

  it('mounts a pipeline that includes middleware', async function () {
    await mount({app, dir: 'test/routes/pipeline/middleware'});

    await pipe({pipeline: app.get.firstCall.args[1], req, res, next});

    assert(next.calledOnce);
    assert(next.firstCall.args[0] instanceof Error);
    assert(next.firstCall.args[0].message === 'MIDDLEWARE');
  });
});


test/mount.js
const assert = require('assert');
const routing = require('../index');
const path = require('path');

describe('Mount', function() {
  it('should mount routes', function() {
    const routes = [{path: '/', get: 'get.js'}];
    let mounted = false;

    const app = {
      get: function(route) {
        mounted = true;
      }
    };

    routing.mount({routes, app});
  });

  it('should throw on invalid permissions format', function() {
    const routes = [{path: '/', patch: path.resolve('./test/mount/permissions/patch.js')}];
    const req = {user: {can: new Function()}};
    let throwed = false;

    const app = {
      patch: function(route, handler) {
        try {
          handler(req);
        } catch (err) {
          throwed = err.message === 'PERMISSIONS_FORMAT_INVALID';
        }
      }
    };

    routing.mount({routes, app});
    assert(throwed);
  });

  it('should check permissions expressed as an object', function() {
    const routes = [{path: '/', post: path.resolve('./test/mount/permissions/post.js')}];
    let checkedCreateUser = false;
    let checkedUpdateFactory = false;

    const req = {
      user: {
        can: function(action, target) {
          if (!checkedCreateUser) checkedCreateUser = action === 'create' && target === 'User';
          if (!checkedUpdateFactory) checkedUpdateFactory = action === 'update' && target === 'Factory';

          return true;
        }
      }
    };

    const app = {post: (route, handler) => handler(req)};

    routing.mount({routes, app});
    assert(checkedCreateUser && checkedUpdateFactory);
  });

  it('should check permissions expressed as a function', function() {
    const routes = [{path: '/', get: path.resolve('./test/mount/permissions/get.js')}];
    let checked = false;

    const res = {locals: {user: 'abc'}};

    const req = {
      user: {
        can: function(action, target) {
          checked = action === 'read' && target === res.locals.user;

          return true;
        }
      }
    };

    const app = {get: (route, handler) => handler(req, res)};

    routing.mount({routes, app});
    assert(checked);
  });
});


test/discover.js
const assert = require('assert');
const routing = require('../index');
const fs = require('fs');

describe('Discover', function() {
  it('should throw on missing path parameter', function() {
    assert.throws(routing.discover, {message: 'PATH_MISSING'});
  });

  it('should throw on invalid path', function() {
    assert.throws(() => routing.discover('abc'), {code: 'ENOENT'});
  });

  it('should ignore empty directories', function() {
    try {
      fs.mkdirSync('test/discover/empty');
    } catch (err) {
      if (err.code !== 'EEXIST') throw err;
    }

    const routes = routing.discover('test/discover/empty');

    assert(routes.length === 0);
  });

  it('should discover valid methods', function() {
    const route = routing.discover('test/discover/valid-methods')[0];

    ['use', 'get', 'post', 'put', 'patch', 'delete'].forEach(method => {
      assert(typeof route[method] !== 'undefined');
    });
  });

  it('should ignore invalid methods', function() {
    const routes = routing.discover('test/discover/invalid-method');

    assert(routes.length === 0);
  });

  it('should output files first', function() {
    const routes = routing.discover('test/discover/file-order');

    assert(routes[0].path === '/' && routes[1].path === '/dir');
  });

  it('should recognize id param routes', function() {
    const route = routing.discover('test/discover/id-param')[0];

    assert(route.path === '/:paramNameId');
  });

  it('should recognize named param routes', function() {
    const route = routing.discover('test/discover/named-param')[0];

    assert(route.path === '/:paramName');
  });

  it('should place param routes at the end', function() {
    const routes = routing.discover('test/discover/param-order');

    assert(routes[0].path === '/z-order' && routes[1].path === '/:paramNameId' && routes[2].path === '/:paramName');
  });
});


lib/mount.js
const fs = require('fs');
const path = require('path');
const errCode = require('err-code');
const httpErrors = require('http-errors');
const ms = require('ms');
const {RateLimiterMemory, RateLimiterRes} = require('rate-limiter-flexible');

/**
 * Place "=..." dirents last in the array.
 * @param {Dirent} a
 * @param {Dirent} b
 * @return {number}
 */
function paramLast(a, b) {
  if (a.name.startsWith('=') && b.name.startsWith('=')) {
    return 0;
  }

  if (a.name.startsWith('=')) {
    return 1;
  }

  if (b.name.startsWith('=')) {
    return -1;
  }

  return 0;
}

/**
 * Create middleware for the rate-limiting feature.
 * @param {string|Object} rateLimit
 * @throws {Error} RATE_LIMIT_INVALID
 * @returns {Function}
 */
function createRateLimitMiddleware(rateLimit) {
  if (typeof rateLimit !== 'string' && typeof rateLimit !== 'object') {
    throw errCode(new Error('The "rateLimit" parameter is not a string or an object'), 'RATE_LIMIT_INVALID');
  }

  if (typeof rateLimit === 'string') {
    rateLimit = {points: 1, duration: ms(rateLimit) / 1000};
  }

  const rateLimiter = new RateLimiterMemory(rateLimit);

  return (req, res, next) => {
    return rateLimiter.consume(req.ip)
      .then(() => next())
      .catch(err => {
        if (err instanceof RateLimiterRes) {
          next(httpErrors(429, err));
        } else {
          next(err);
        }
      });
  };
}

/**
 * Create middleware for the authorization feature.
 * @param {Function} authorize
 * @throws {Error} AUTHORIZE_INVALID
 * @returns {Function}
 */
function createAuthorizeMiddleware(authorize) {
  if (typeof authorize !== 'function') {
    throw errCode(new Error('The "authorize" parameter is not a function'), 'AUTHORIZE_INVALID');
  }

  return (req, res, next) => {
    return Promise.resolve(authorize(req, res))
      .then(isAuthorized => {
        if (isAuthorized !== true) {
          next(httpErrors(403));
        } else {
          next();
        }
      })
      .catch(next);
  };
}

/**
 * Create middleware for the request handler.
 * @param {Function} handle
 * @throws {Error} HANDLE_INVALID
 * @returns {Function}
 */
function createHandleMiddleware(handle) {
  if (typeof handle !== 'function') {
    throw errCode(new Error('The "handle" parameter is not a function'), 'HANDLE_INVALID');
  }

  return (req, res, next) => Promise.resolve(handle(req, res)).catch(next);
}

/**
 * Create a sequence of middleware function to process a request.
 * @param {string|Object} [rateLimit]
 * @param {Function} [authorize]
 * @param {Function|Function[]} [middleware]
 * @param {Function} handle
 * @returns {Function[]}
 */
function assemblePipeline({rateLimit, authorize, middleware, handle} = {}) {
  const pipeline = [];

  if (typeof rateLimit !== 'undefined') {
    pipeline.push(createRateLimitMiddleware(rateLimit));
  }

  if (typeof authorize !== 'undefined') {
    pipeline.push(createAuthorizeMiddleware(authorize));
  }

  pipeline.push(middleware);
  pipeline.push(createHandleMiddleware(handle));

  return pipeline.flat().filter(element => typeof element === 'function');
}

/**
 *
 * @param {string} dir
 * @param {Object} app
 * @param {string} [route]
 * @param {Array} [mounted=[]]
 * @throws {Error} DIR_INVALID
 * @throws {Error} ROOT_INVALID
 */
function mount({dir, app, route = '/', mounted = []} = {}) {
  if (typeof dir !== 'string') {
    throw errCode(new Error('The "dir" parameter is not a string'), 'DIR_INVALID');
  }

  if (dir.trim() === '') {
    throw errCode(new Error('The "dir" parameter can not be an empty string'), 'DIR_INVALID');
  }

  if (typeof route !== 'string') {
    throw errCode(new Error('The "route" parameter is not a string'), 'ROOT_INVALID');
  }

  if (route.trim() === '') {
    throw errCode(new Error('The "route" parameter can not be an empty string'), 'ROOT_INVALID');
  }

  const dirents = fs.readdirSync(dir, {withFileTypes: true});

  dirents
    .filter(dirent => dirent.isFile())
    .forEach(dirent => {
      const direntPath = path.resolve(path.join(dir, dirent.name));
      const source = require(direntPath);
      const pipeline = assemblePipeline(source);
      const method = path.basename(dirent.name, '.js');

      mounted.push({route, method, source, pipeline});
      app[method](route, pipeline);
    });

  // routes with parameters must be mounted the last for other routes to be reachable
  dirents
    .filter(dirent => dirent.isDirectory()).sort(paramLast)
    .forEach(dirent => {
      let direntBranch;

      if (dirent.name === '=') {
        direntBranch = '*';
      } else if (dirent.name.startsWith('=')) {
        direntBranch = dirent.name.replace('=', ':').replace(/-(\w)/g, (match, p1) => p1.toUpperCase());
      } else {
        direntBranch = dirent.name;
      }

      mount({
        app,
        mounted,
        dir: path.resolve(path.join(dir, dirent.name)),
        // names equal to "=" are used to represent "*" wildcard
        // names starting with the "=" character are used to represent parameters
        route: path.posix.join(route, direntBranch)
      });
    });

  return mounted;
}

module.exports = mount;


index.js
import fs from 'fs';
import path from 'path';
import errCode from 'err-code';
import httpErrors from 'http-errors';
import ms from 'ms';
import {RateLimiterMemory, RateLimiterRes} from 'rate-limiter-flexible';

/**
 * Place "=..." dirents last in the array.
 * @param {Dirent} a
 * @param {Dirent} b
 * @return {number}
 */
function paramLast(a, b) {
  if (a.name.startsWith('=') && b.name.startsWith('=')) {
    return 0;
  }

  if (a.name.startsWith('=')) {
    return 1;
  }

  if (b.name.startsWith('=')) {
    return -1;
  }

  return 0;
}

/**
 * Create middleware for the rate-limiting feature.
 * @param {string|Object} rateLimit
 * @throws {Error} RATE_LIMIT_INVALID
 * @returns {Function}
 */
function createRateLimitMiddleware(rateLimit) {
  if (typeof rateLimit !== 'string' && typeof rateLimit !== 'object') {
    throw errCode(new Error('The "rateLimit" parameter is not a string or an object'), 'RATE_LIMIT_INVALID');
  }

  if (typeof rateLimit === 'string') {
    rateLimit = {points: 1, duration: ms(rateLimit) / 1000};
  }

  const rateLimiter = new RateLimiterMemory(rateLimit);

  return (req, res, next) => {
    return rateLimiter.consume(req.ip)
      .then(() => next())
      .catch(err => {
        if (err instanceof RateLimiterRes) {
          next(httpErrors(429, err));
        } else {
          next(err);
        }
      });
  };
}

/**
 * Create middleware for the authorization feature.
 * @param {Function} authorize
 * @throws {Error} AUTHORIZE_INVALID
 * @returns {Function}
 */
function createAuthorizeMiddleware(authorize) {
  if (typeof authorize !== 'function') {
    throw errCode(new Error('The "authorize" parameter is not a function'), 'AUTHORIZE_INVALID');
  }

  return (req, res, next) => {
    return Promise.resolve(authorize(req, res))
      .then(isAuthorized => {
        if (isAuthorized !== true) {
          next(httpErrors(403));
        } else {
          next();
        }
      })
      .catch(next);
  };
}

/**
 * Create middleware for the request handler.
 * @param {Function} handle
 * @throws {Error} HANDLE_INVALID
 * @returns {Function}
 */
function createHandleMiddleware(handle) {
  if (typeof handle !== 'function') {
    throw errCode(new Error('The "handle" parameter is not a function'), 'HANDLE_INVALID');
  }

  return (req, res, next) => Promise.resolve(handle(req, res)).catch(next);
}

/**
 * Create a sequence of middleware function to process a request.
 * @param {string|Object} [rateLimit]
 * @param {Function} [authorize]
 * @param {Function|Function[]} [middleware]
 * @param {Function} handle
 * @returns {Function[]}
 */
function assemblePipeline({rateLimit, authorize, middleware, handle} = {}) {
  const pipeline = [];

  if (typeof rateLimit !== 'undefined') {
    pipeline.push(createRateLimitMiddleware(rateLimit));
  }

  if (typeof authorize !== 'undefined') {
    pipeline.push(createAuthorizeMiddleware(authorize));
  }

  pipeline.push(middleware);
  pipeline.push(createHandleMiddleware(handle));

  return pipeline.flat().filter(element => typeof element === 'function');
}

/**
 * Find route handlers in the directory.
 * @param {string} dir
 * @param {string} [route]
 * @returns {Promise<*[]>}
 */
async function discover({dir, route = '/'} = {}) {
  if (typeof dir !== 'string') {
    throw errCode(new Error('The "dir" parameter is not a string'), 'DIR_INVALID');
  }

  if (dir.trim() === '') {
    throw errCode(new Error('The "dir" parameter can not be an empty string'), 'DIR_INVALID');
  }

  if (typeof route !== 'string') {
    throw errCode(new Error('The "route" parameter is not a string'), 'ROOT_INVALID');
  }

  if (route.trim() === '') {
    throw errCode(new Error('The "route" parameter can not be an empty string'), 'ROOT_INVALID');
  }

  const handlers = [];
  const dirents = fs.readdirSync(dir, {withFileTypes: true});

  for (const dirent of dirents.filter(dirent => dirent.isFile())) {
    const direntPath = path.resolve(path.join(dir, dirent.name));
    const module = await import(direntPath);
    const pipeline = assemblePipeline(module.default);
    const method = path.basename(dirent.name, '.js');

    handlers.push({route, method, module, pipeline});
  }

  // routes with parameters must be mounted the last for other routes to be reachable
  for (const dirent of dirents.filter(dirent => dirent.isDirectory()).sort(paramLast)) {
    let direntBranch;

    if (dirent.name === '=') {
      // names equal to "=" are used to represent "*" wildcard
      direntBranch = '*';
    } else if (dirent.name.startsWith('=')) {
      // names starting with the "=" character are used to represent parameters
      direntBranch = dirent.name.replace('=', ':').replace(/-(\w)/g, (match, p1) => p1.toUpperCase());
    } else {
      direntBranch = dirent.name;
    }

    const direntDir = path.resolve(path.join(dir, dirent.name));
    const direntRoute = path.posix.join(route, direntBranch);

    (await discover({dir: direntDir, route: direntRoute})).forEach(handler => handlers.push(handler));
  }

  return handlers;
}

/**
 * Mount route handlers to the app.
 * @param {string} [dir]
 * @param {Array} [handlers]
 * @param {Object} app
 * @param {string} [route]
 * @returns {Promise<*[]>}
 */
async function mount({dir, handlers, app, route = '/'} = {}) {
  if (typeof dir !== 'string' && !Array.isArray(handlers)) {
    throw errCode(new Error('Missing "dir" (string) or "handlers" (array) parameter'), 'MODULE_INVALID');
  }

  if (!Array.isArray(handlers)) {
    handlers = await discover({dir, route});
  }

  handlers.forEach(({route, method, pipeline}) => app[method](route, pipeline));

  return handlers;
}

export {discover, mount};
